compiler/rustc_resolve/src/imports.rs

```diff
@@ -41,6 +42,93 @@ use crate::{
 
 type Res = def::Res<NodeId>;
 
+pub(crate) struct ImportResolver<'r, 'ra, 'tcx> {
+    r: CmResolver<'r, 'ra, 'tcx>, // always immutable
+    outputs: ImportResolutionOutputs<'ra>,
+}
+
+#[derive(Default)]
+struct ImportResolutionOutputs<'ra> {
+    prelude: Option<Module<'ra>>,
+    indeterminate_imports: Vec<Import<'ra>>,
+    determined_imports: Vec<Import<'ra>>,
+    module_glob_importers: FxIndexMap<Module<'ra>, Vec<Import<'ra>>>,
+    glob_import_bindings:
+        Vec<(Module<'ra>, BindingKey, NameBinding<'ra>, bool /* warn_ambiguity */)>,
+    glob_path_res: Vec<(NodeId, PartialRes)>,
+    single_import_bindings: PerNS<Vec<(Module<'ra>, Import<'ra>, PendingBinding<'ra>)>>,
+}
```

Comment:
I suggest restructuring this structure like this:
```rust
struct ImportResolutionOutputs {
  indeterminate_imports: Vec<Import>,
  determined_imports: Vec<(Import, SideEffect)>,
}
```
where `SideEffect` is a set of changes that need to be applied for that specific import.

Then side effects will be applied in the same order in which the import are resolved.
It will avoid having superfluous differences with the old algorithm, in emitted diagnostics in particular, and should make debugging the remaining differences easier.